### PR TITLE
Manalink: Send mana to anyone via link

### PR DESCRIPTION
**Pull request by akrolsmir**
See proposal: https://manifoldmarkets.notion.site/External-Bounty-Links-Manalinks-b6a1f196d04646bfb05ecb90d9f023e3

### CODE REVIEW

**Comment by Gurkenglas** [comment]
> If a new user signs up to claim this, sender + new user both get a bonus e.g. M$ 200

That recovers some of the ick. Consider letting the visitor using the link bet without having made an account. The point was that you don't need to fear account spam to farm signup bonuses when all the mana comes from users.

File: common/manalink.ts
Comment (jahooma) [inline common/manalink.ts:7]: How would we cross reference txns to mana links? Maybe the txn will have the id / slug of the manalink in a data field.

**Review by jahooma — COMMENTED** [review]
Cool idea!

File: common/manalink.ts
Comment (akrolsmir) [reply common/manalink.ts:1]: Eh, I was hoping to coin a new word "Manalink". It's an internal name anyways

**Comment by akrolsmir** [comment]
@Gurkenglas - The idea to create an anonymous account is pretty interesting, not one I'd considered before.

But not sure how much work that'd be to handle, plus we kinda do want people to be signing up... Otherwise we're open to e.g. assassination markets where someone can move large amounts of M$ and there's no way for us to trace it.

Re: Expiration dates: from the manalink creator side they also don't want to have this possibility of the link getting claimed eg 3 years into the future. Think of it like the close date on a market.

File: common/manalink.ts
Comment (akrolsmir) [reply common/manalink.ts:4]: The document key is going to be the slug/id as well!

File: web/components/manalink-card.tsx
Comment (sipec) [inline web/components/manalink-card.tsx:35]: copy: "Unlimited use" ?

File: web/pages/link.tsx
Comment (sipec) [inline web/pages/link.tsx]: delete this.

personally, I like to type `const p = (x: any) => (console.log(x), x)` at the top of files I'm debugging, then I can just use `p` to print expressions in the middle of statements without having to copy them onto a new line.

**Review by sipec — COMMENTED** [review]
Just reviewed mqp's changes. Seem pretty good overall - I especially like the LinksTable refactor.

File: functions/src/transact.ts
Comment (sipec) [inline functions/src/transact.ts:74]: I love this simplification!

File: web/lib/firebase/txns.ts
Comment (sipec) [inline web/lib/firebase/txns.ts:68]: maybe `orderBy([...fromTxns, ...toTxns], ['createdTime'], ['desc']`

File: web/pages/link.tsx
Comment (sipec) [inline web/pages/link.tsx:73]: maybe `createManaLink({ fromId: user.id, ...newManalink })` ?

File: web/pages/link.tsx
Comment (sipec) [inline web/pages/link.tsx:87]: [Not important] Since all the ways of updating a mana link follow this structure, a neat pattern is to replace the useState with a simple useReducer:
```
const [newManaLink, updateNewManaLink] = useReducer((m, update) => ({ ...m, ...update }), (default value here))
```
then on these lines you just update like `updateNewManaLink({ amount: parseInt(e.target.value) })`

**Review by sipec — APPROVED** [review]
... and now I've reviewed the rest of it.
I would like another pair of eyes on the firebase stuff though

File: web/pages/link.tsx
Comment (mqp) [reply web/pages/link.tsx]: There's actually another field I don't want in there, so this isn't quite right.

**Comment by mqp** [comment]
Thanks for the feedback @sipec. This is officially good to merge in my book, it doesn't look totally crappy, I tested stuff, and I will merge at my earliest convenience when I can take care of updating security rules & functions & etc.

**Review by akrolsmir — COMMENTED** [review]
Great work, Marshall! I think the main missing thing from a product perspective is the ability to expire a link - did we decide this wasn't as important for v0?

Other things I'm jotting down for follow-up work:
**Create tab**
- [ ] Expires At => Expires in dropdown that has "1 day / 1 week / never" options ala Discord 
image

- [ ] Maybe also the same for max uses? less sure 
image

**Links tab**
- [ ] Copy text tweaks: "Claimant" to "Claimed by", "Outstanding links" to "Unclaimed links"
- [ ] Created but claimed links should be visible somewhere (eg a new "Claimed links" tab)

**Claim page**
- [ ] Narrow the width of the preview, to match the claim page
- [ ] If there's a message included, also display who the sender is on the claim page. Maybe include their avatar on the bottom right before the message eg "😄 Austin: for fixing our graphs"
- [ ] Claimed links should be more obviously not claimable - disable/hide claim button; change text

**Comment by mqp** [comment]
> I think the main missing thing from a product perspective is the ability to expire or delete your own link - did we decide this wasn't as important for v0?

I think it's important to get this out so that the branch doesn't keep going out of date and so that David can reimburse survey participants. Perhaps we should add more features for managing your own links before we expose the link creation UI in the main navigation.

**Pull request merged**
